### ISSUE TITLE
ESR 45: Default preference files fail to load

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/mozilla-os2#198]
The debug build shows that the new ESR 45 build fails to load some preferences. Here's the respective output:
```
[3354] WARNING: malformed pref file: file D:/Coding/mozilla/master/modules/libpref/prefread.cpp, line 351
[3354] ###!!! ASSERTION: Default pref file not parsed successfully.: 'Error', file D:/Coding/mozilla/master/modules/libpref/Preferences.cpp, line 1116
[3354] WARNING: malformed pref file: file D:/Coding/mozilla/master/modules/libpref/prefread.cpp, line 351
[3354] ###!!! ASSERTION: Default pref file not parsed successfully.: 'Error', file D:/Coding/mozilla/master/modules/libpref/Preferences.cpp, line 1116
[3354] WARNING: Error parsing application default preferences.: file D:/Coding/mozilla/master/modules/libpref/Preferences.cpp, line 1287
[3354] WARNING: malformed pref file: file D:/Coding/mozilla/master/modules/libpref/prefread.cpp, line 351
[3354] ###!!! ASSERTION: Default pref file not parsed successfully.: 'Error', file D:/Coding/mozilla/master/modules/libpref/Preferences.cpp, line 1116
JavaScript error: file:///D:/CODING/MOZILLA/MASTER-BUILD-DEBUG/DIST/BIN/components/Weave.js, line 1: SyntaxError: unterminated comment
[3354] WARNING: LoadModule failed for module 'file:///D:/CODING/MOZILLA/MASTER-BUILD-DEBUG/DIST/BIN/components/Weave.js': file D:/Coding/mozilla/master/xpcom/components/nsComponentManager.cpp, line 915
[3354] WARNING: Cannot create startup observer : service,@mozilla.org/weave/service;1: file D:/Coding/mozilla/master/embedding/components/appstartup/nsAppStartupNotifier.cpp, line 81
JavaScript error: resource://gre/modules/XPCOMUtils.jsm, line 1: SyntaxError: unterminated comment
[3354] WARNING: LoadModule failed for module 'file:///D:/CODING/MOZILLA/MASTER-BUILD-DEBUG/DIST/BIN/browser/components/nsBrowserGlue.js': file D:/Coding/mozilla/master/xpcom/components/nsComponentManager.cpp, line 915
[3354] WARNING: Cannot create startup observer : service,@mozilla.org/browser/browserglue;1: file D:/Coding/mozilla/master/embedding/components/appstartup/nsAppStartupNotifier.cpp, line 81
JavaScript error: file:///D:/CODING/MOZILLA/MASTER-BUILD-DEBUG/DIST/BIN/components/addonManager.js, line 1: SyntaxError: unterminated comment
[3354] WARNING: LoadModule failed for module 'file:///D:/CODING/MOZILLA/MASTER-BUILD-DEBUG/DIST/BIN/components/addonManager.js': file D:/Coding/mozilla/master/xpcom/components/nsComponentManager.cpp, line 915
[3354] WARNING: Failed to create Addons Manager.: file D:/Coding/mozilla/master/toolkit/xre/nsXREDirProvider.cpp, line 888
JavaScript error: file:///D:/CODING/MOZILLA/MASTER-BUILD-DEBUG/DIST/BIN/components/nsBlocklistService.js, line 3: SyntaxError: unterminated comment
[3354] WARNING: LoadModule failed for module 'file:///D:/CODING/MOZILLA/MASTER-BUILD-DEBUG/DIST/BIN/components/nsBlocklistService.js': file D:/Coding/mozilla/master/xpcom/components/nsComponentManager.cpp, line 915
```

**Comment by dmik** [comment]
My guess is that these .js files from `components/` fail to load because they are symlinks now (used to be regular files in our ESR 38 builds). And the JS engine somehow seems to use a DOS API directly to load these files instead of kLIBC, hence reading the symlink and parsing the actual target leads to reading a regular file and parsing a file name it contains. Hence the parse errors.

**Comment by dmik** [comment]
Well, these files are read by `nsFileInputStream` which is NSPR-based, so  I will try to hack some bits around there.

**Comment by dmik** [comment]
Well, I got rid of these warnings in libpref by fixing NSPR, but the JavaScript ones are still there. I don't get what were the libperf ones about then.

**Comment by dmik** [comment]
Turns out that the remaining problems are NSPR too. The `mozJSComponentLoader` class uses it to load .js file contents as well. However the failure was much more tricky there: The `nsLocalFileOS2` implementation uses `PR_GetFileInfo64` which does `stat` for most fields but the file size; for the latter it would call `DosQueryPathInfo` directly which returned the size of the symlink of course, not the size of the target as `stat` does. Hence the `unterminated comment` JS parser error. I will have to fix NSPR to use kLIBC calls for all places involving file names.

I created http://trac.netlabs.org/ports/ticket/153 for that.
